design/one-pager-strongly-typed-class.md

```diff
@@ -0,0 +1,218 @@
+
+## Goals
+
+* Enable schema documentation and validation for resource classes.
+* Continue to support configuring a default resource class to be set upon resource claims that do not specify a class.
+* Assume that the universe of possible default resource class kinds for a specific claim kind will grow arbitrarily, so a default class controller must be able to be made aware of new resource class kinds. This goal will become specifically important as we begin to [separate providers](https://github.com/crossplaneio/crossplane/issues/531) from the core Crossplane project in the form of infrastructure stacks. 
+
+## Proposal
+
+Unfortunately, strongly typed resource classes make it much more difficult to identify default resource classes. Currently, there is one default class controller per claim kind that watches for creation of their specified claim kind, then lists objects of `kind: ResourceClass` that are labeled as default for the claim kind. With an unknown number of possible kinds of resource classes that could be installed and serve as default for a claim `kind`, a default class controller can no longer list a single `ResourceClass` kind.
+
+One solution would instead be to map resource classes one-to-one with claim kinds (i.e. `MySQLInstanceClass`), but that results in the same issue of varying required parameters that is currently being experienced.
+
+In order to satisfy both the requirements of strong typing and default resource classes, a new custom resource kind, `Defaulter`, can be implemented to specify when a specific resource class kind can serve as default for a given claim kind. The workflow would proceed as follows:
+
+1. An admin installs Crossplane, which specifies the portable claim kinds that are available as part of the core Crossplane.
+2. An admin installs an infrastructure stack (i.e. a set of CRDs and controllers for a provider's managed service offerings).
+3. An admin creates an instance of an installed resource class CRD (i.e. `kind: RDSInstanceClass`). It could look as follows:
+
+```yaml
+apiVersion: database.aws.crossplane.io/v1alpha1
+kind: RDSInstanceClass
+metadata:
+  name: standard-mysql
+  namespace: crossplane-system
+spec:
+  class: db.t2.small
+  masterUsername: masteruser
+  securityGroups: "sg-ab1cdefg,sg-05adsfkaj1ksdjak"
+  size: "20"
+  provisioner: rdsinstance.database.aws.crossplane.io/v1alpha1
+providerRef:
+  name: aws-provider
+reclaimPolicy: Delete
+```
+
+4. An admin creates a new `Defaulter` object, specifying that the previously created resource class (`kind: RDSInstanceClass`) should serve as default for a given claim kind `mysqlinstance.storage.crossplane.io`. A `Defaulter` could look as follows:
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: Defaulter
+metadata:
+  name: mysql-aws-defaulter
+  namespace: crossplane-system
```

Comment:
It's possible, even likely, that managed resources and resource classes will become cluster scoped resources in the near future. Resource claims will remain namespaced. I wonder whether defaulters should be cluster scoped or namespace scoped in such a world?

I actually think we should have both, ala [`RoleBinding` and `ClusterRoleBinding`](https://kubernetes.io/docs/reference/access-authn-authz/rbac/). I could imagine that a cluster operator might want to set defaults at the cluster scope, but also allow app operators to set defaults at the namespace scope.

Reply:
I think this is the right course of action, and thank you for linking the relevant docs! With this in mind, I believe that we will have two CRD's per claim kind (i.e. `MySQLInstancePolicy` and `MySQLInstanceClusterPolicy`).